src/devices/cpu/pic16x8x/pic16x8x.cpp

```diff
@@ -0,0 +1,1443 @@
+ The address depends on the device used.
+Devices with 36 bytes ends at 2Fh, devices with
+68 bytes ends at 4Fh.*/
+
+void pic16x8x_device::core_regs(address_map &map, u8 mirror)
+{
+	map(0x00, 0x00).noprw().mirror(mirror); // Not an actual register, reading indirectly (FSR=0) returns 0
+	map(0x01, 0x01).rw(FUNC(pic16x8x_device::tmr0_r), FUNC(pic16x8x_device::tmr0_w));
+	map(0x02, 0x02).rw(FUNC(pic16x8x_device::pcl_r), FUNC(pic16x8x_device::pcl_w)).mirror(mirror);
+	map(0x03, 0x03).rw(FUNC(pic16x8x_device::status_r), FUNC(pic16x8x_device::status_w)).mirror(mirror);
+	map(0x04, 0x04).rw(FUNC(pic16x8x_device::fsr_r), FUNC(pic16x8x_device::fsr_w)).mirror(mirror);
+	map(0x05, 0x05).rw(FUNC(pic16x8x_device::porta_r), FUNC(pic16x8x_device::porta_w));
+	map(0x06, 0x06).rw(FUNC(pic16x8x_device::portb_r), FUNC(pic16x8x_device::portb_w));
+	map(0x07, 0x07).noprw().mirror(mirror); // Not an actual register, returns 0
+	map(0x08, 0x08).rw(FUNC(pic16x8x_device::eedata_r), FUNC(pic16x8x_device::eedata_w));
+	map(0x09, 0x09).rw(FUNC(pic16x8x_device::eeadr_r), FUNC(pic16x8x_device::eeadr_w));
+	map(0x0a, 0x0a).rw(FUNC(pic16x8x_device::pclath_r), FUNC(pic16x8x_device::pclath_w)).mirror(mirror);
+	map(0x0b, 0x0b).rw(FUNC(pic16x8x_device::intcon_r), FUNC(pic16x8x_device::intcon_w)).mirror(mirror);
+	map(0x81, 0x81).rw(FUNC(pic16x8x_device::option_r), FUNC(pic16x8x_device::option_w));
+	map(0x85, 0x85).rw(FUNC(pic16x8x_device::trisa_r), FUNC(pic16x8x_device::trisa_w));
+	map(0x86, 0x86).rw(FUNC(pic16x8x_device::trisb_r), FUNC(pic16x8x_device::trisb_w));
+	map(0x88, 0x88).rw(FUNC(pic16x8x_device::eecon1_r), FUNC(pic16x8x_device::eecon1_w));
+	map(0x89, 0x89).rw(FUNC(pic16x8x_device::eecon2_r), FUNC(pic16x8x_device::eecon2_w));
+}
+
+void pic16x8x_device::ram_6(address_map &map)
+{
+	// 0x00 - 0x0b SFR's Bank 0
+	// 0x0c - 0x2f GPR's
+	// 0x80 - 0x8b SFR's Bank 1
+	// 0x8c - 0xaf GPR Mirrored to 0x0c - 0x2f
+	core_regs(map, 0x80);
+	map(0x0c, 0x2f).ram().mirror(0x80);
+}
+
+void pic16x8x_device::ram_7(address_map &map)
+{
+	// 0x00 - 0x0b SFR's Bank 0
+	// 0x0c - 0x4f GPR's
+	// 0x80 - 0x8b SFR's Bank 1
+	// 0x8c - 0xcf GPR Mirrored to 0x0c - 0x4f
+	core_regs(map, 0x80);
+	map(0x0c, 0x4f).ram().mirror(0x80);
+}
+
+pic16x8x_device::pic16x8x_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, u32 clock, int program_width, uint16_t eeprom_size, int pic_model)
+	: cpu_device(mconfig, type, tag, owner, clock)
+	, device_nvram_interface(mconfig, *this)
+	, m_region(*this, DEVICE_SELF)
+	, m_program_config("program", ENDIANNESS_LITTLE, 16, program_width, -1,
+			((program_width == 9) ? address_map_constructor(FUNC(pic16x8x_device::rom_9), this) :  address_map_constructor(FUNC(pic16x8x_device::rom_10),this)))
+	, m_data_config("data", ENDIANNESS_LITTLE, 8, 8, 0,
+			((pic_model == 0x83) ? address_map_constructor(FUNC(pic16x8x_device::ram_6), this) :
+			address_map_constructor(FUNC(pic16x8x_device::ram_7), this)))

```

Comment:
It's probably better just to pass the address map constructor here and eliminate the artificial `pic_model` constants.

Reply:
As I can undertand, I guess I should split  the constructor  like pic17c4x code does. If I'm right please confirm.

Reply:
Done. Deleted pic_model and and use program_width as selector.

Reply:
Second revision. Fixed.
 Added new base classes and instances for 16x83 and 16x84 to define properly the address maps.